### ISSUE TITLE
Support for hiding function bodies at use sites, as a dual to `reveal`

### ISSUE THREAD

**Issue by atomb** [dafny-lang/dafny#3884]
### Summary

As a dual to the `reveal` command, provide a `hide` or `unreveal` command that would make the body of the given function unavailable to the verifier.

As a possible complement, a new attribute on definitions, maybe `{:auto_reveal bool}` could control the default treatment of functions. A definition with `{:auto_reveal false}` would treat all functions as `opaque`, even if they weren't marked that way. In a sense, this would switch Dafny from "as automatic as possible" mode into "more like an interactive theorem prover" mode.

### Background and Motivation

Currently, functions in Dafny are transparent by default, and all bodies are available to the verifier. This can make verification very automatic in many cases, but sometimes can bog down the verifier. Making a function `opaque` can help avoid this issue by hiding the body until it is requested with the `reveal` command.

However, in a large development, making a function `opaque` can cause other previously-successful proofs to fail, requiring a scattering of `reveal` commands throughout a program.

### Proposed Feature

This feature would introduce one additional keyword and possibly one additional attribute.

The statement `hide F();` or `unreveal F();` (name still TBD) would instruct the verifier to not consider the body of `F` during verification. If `F` was already declared `opaque`, it would have no effect.

The potential attribute, the presence of `{:auto_reveal bool}` on a definition would control whether function bodies are available by default when verifying that function. If `{:auto_reveal true}` is present, Dafny would behave as it does now. If `{:auto_reveal false}` is present, Dafny would behave as if every function in the program was declared `opaque`, but only during the verification of that particular definition.

### Alternatives

The key alternative is to leave Dafny as-is, encourage Dafny programmers to make most functions `opaque`, and to sprinkle `reveal` commands throughout, as needed.

The key downsides of this approach are:
* New programs will more verbose than they need to be.
* Some programs will require many `reveal` statements when a smaller number `hide` statements would be equally effective.
* Large, existing programs in which extra opacity is necessary for verification performance may be hard to modify, because making a given function opaque could require a large number of new `reveal` statements throughout the program.

**Comment by MikaelMayer** [comment]
With https://github.com/dafny-lang/dafny/pull/3779, it will become possible to implement such a feature by changing at a call site the use of a _global reveal variable_ to a _local reveal variable_ that is set to true by default (if {:auto-reveal true}), or false (if {:autoreveal false}). Revealing is a matter of assigning "true" to it, and hiding is a matter of "havoc" it.

**Comment by atomb** [comment]
Based on @MikaelMayer's changes in #3779, I think the following is one possible approach to the design.

We can abstractly consider every function to be in one of three categories:
* Opaque (if it is marked with the `opaque` keyword or `{:opaque}` modifier
* Hidable (if it can appear as the argument to `hide`)
* Normal

The reason to potentially distinguish "hidable" from "normal" is efficiency, as I'll go into below, along with the criteria for distinguishing between them.

When generating Boogie, generate top-level constants for containing the definitions of all functions, as done currently. Opaque and hidable functions will have reveal parameters. Other functions will not.

Add boolean "reveal" constants for opaque functions to indicate whether they're revealed. Skip doing this for hidable functions.

In the body of each definition, add a boolean "reveal" variable for each hidable function, set to "true" by default. The `hide` statement will be translated into an assignment setting one of these to `false`. Having a distinction between hidable and non-hidable functions reduces the number of variables added to every definition.

When calling a hidable function, pass in the local "reveal" variable instead of a global one (which might not be a change if they have the same name).

To distinguish between hidable and normal functions, we could do an initial pass over the whole program to identify any function that appears as the argument of a `hide` statement.

A simplification would be to say that every function takes a "reveal" parameter and that the variables controlling this parameter are all local, declared for all functions at the beginning of every definition. Optimization by removing unused locals would reduce the noise. I don't know whether adding an extra parameter to every function would have a performance impact.

**Comment by atomb** [comment]
As an experiment, I'm going to implement and test the performance of a change that:
* Makes all functions take a "reveal" argument
* Makes each "reveal" variable a separate local variable in each function

If this doesn't hurt performance, it'll make the above a lot simpler.

**Comment by keyboardDrummer** [comment]
Is there a chance you could an provide an example of a situation in which you are calling a function in two locations, where in one location you want it to be revealed, and in the other you don't?

I imagine that any opaque function should have an ensures clause, otherwise you have no idea of what it does without revealing it, and if you'd always have to reveal it then there'd be no sense in making it opaque.

But then, if you have an ensures clause on your function, what are the situations in which you still want to reveal it?

**Comment by MikaelMayer** [comment]
> My assumption is that if you're using a function, then you need to know something about it other than its types, otherwise you won't be able to do fully verify the code that's calling it.

You can think of the following example, which only assume that the values returned are using the function itself without the need to inspect what the function definition is.

```dafny
function test(i: int): int {
  // Extremely complex facts here
  i + 2
}

method EnsuresMapTest(s: seq<int>) returns (t: seq<int>)
  ensures |t| == |s|
  ensures forall i | 0 <= i < |s| :: t[i] == test(s[i])
{
  t := [];
  while |t| < |s|
    invariant |t| <= |s|
    invariant forall i | 0 <= i < |t| :: t[i] == test(s[i])
  {
    t := t + [test(s[|t|])];
  }
}
```

In the case above, if you moreover imagined:
* That test() is actually calling many other mutually recursive complex functions
* That you have other invariants to assume and to prove in your while loop
you can imagine how the verifier could get stuck sometimes if the invariants you want to prove don't depend on the definition of the function.

More generally, code that has a call to an opaque function don't necessarily need to inspect the content of the function, especially, if lemmas took care of asserting the right properties for this function.

**Comment by keyboardDrummer** [comment]
> You can think of the following example, which only assume that the values returned are using the function itself without the need to inspect what the function definition is.
> 
> ```
> function test(i: int): int {
>   // Extremely complex facts here
>   i + 2
> }
> 
> method EnsuresMapTest(s: seq<int>) returns (t: seq<int>)
>   ensures |t| == |s|
>   ensures forall i | 0 <= i < |s| :: t[i] == test(s[i])
> {
>   t := [];
>   while |t| < |s|
>     invariant |t| <= |s|
>     invariant forall i | 0 <= i < |t| :: t[i] == test(s[i])
>   {
>     t := t + [test(s[|t|])];
>   }
> }
> ```

Thanks, this seems like at least a rare case though. Are there other scenarios than maps on various datatypes, where you want to call a function but not understand its result?

> More generally, code that has a call to an opaque function don't necessarily need to inspect the content of the function, especially, if lemmas took care of asserting the right properties for this function.

I think the lemmas would then be similar to ensures clauses on the opaque function.

**Comment by keyboardDrummer** [comment]
> As a more concrete use case, consider [this function](https://github.com/dafny-lang/libraries/blob/master/src/JSON/Utils/Cursors.dfy#L16-L18) from the JSON code in the libraries repo. It uses sequence concatenation on the result of an arbitrary function, which is a bit tricky to reason about. However, most of the time, the verifier knows this predicate is true from one location and just needs to propagate that to another location by name, never needing to care how that predicate is defined.

Can you not automatically hide it when you get the predicate for free, such as when calling a function that has the predicate in its ensures clause, but automatically reveal it when you're the one who has to prove it?

**Comment by MikaelMayer** [comment]
Solved by #4342
